src/package-tree.ts

```diff
@@ -1,53 +1,116 @@
-  // step 2: add POI to PackageTree Object
+/**
+ * Takes in the root directory of a project and returns returns a PackageTree
+ */
+export async function generatePackageTree(rootDir: string, customReadFilep: ReadFileP): Promise<PackageTree<undefined>> {
+  
+  customReadFilep = customReadFilep || readFilep;
+
+  try{
+    // Step 0: read in package.json and package-lock.json
+    const pjsonPath = path.join(rootDir, 'package.json');
+    const pjson = await customReadFilep(pjsonPath, 'utf8');
+    const pjsonObj = JSON.parse(pjson);
+    
+    const pjsonLockPath = path.join(rootDir, 'package-lock.json');
+    const pjsonLock = await customReadFilep(pjsonLockPath, 'utf8');
+    const pjsonLockObj = JSON.parse(pjsonLock);
+    
+    // Step 1: Get the top level dependencies from pjson
+    const dependencies = {... pjsonObj.dependencies, ...pjsonObj.devDependencies};
+
+    // Step 2: Look at package-lock.json to find dependendencies
+    const result =  await getPackageTreeFromDependencyList(dependencies, pjsonLockObj);
+
+    const treeHead: PackageTree<undefined> = {"rootPackageName": pjsonObj.name,
+      "version": pjsonObj.version,
+      "data": undefined,
+      "dependencies": result}; 
+    
+      return treeHead;
+  }catch(error){
+    throw error;
+  }
+
 }
 
-function getPackagePOIList(package: PackageTree): PointOfInterest[] {
-  // TODO:
-  // calls getPointsOfInterest for each file in package
-  throw new Error('not implemented');
+/**
+ * Takes in a list of depndencies and returns a constructed PackageTree
+ * @param packageLock the package-lock.json file of the root project
+ */
+export async function getPackageTreeFromDependencyList(dependencies: {}, packageLock: {}): Promise<Array<PackageTree<undefined>>|undefined>{
+// Step 1: For each dependency create a PackageTree obj with the name and version fields populated
+if(dependencies){
+  return undefined;
 }
+const dependencyArr: Array<[string,string]> = Object.entries(dependencies);
+const packageTreeArr: Array<PackageTree<undefined>> = [];
+ dependencyArr.forEach((element: [string, string]) => {
+   const pkgTreeObj: PackageTree<undefined> = { "rootPackageName": element[0],
+                                                "version": element[1],
+                                                "data": undefined,
+                                                "dependencies": []
+                                                }; 
+                                                
+    packageTreeArr.push(pkgTreeObj);
+ });
 
-// Gets Points of Interest for a single file
-function getPointsOfInterest(
-    contents: string, fileName: string,
-    functionArray: POIDetector[]): PointOfInterest[] {
-  // TODO:
-  throw new Error('not implemented');
+// Step 2: For each Package Tree obj get the dependencies
+const m: Array<Promise<void>> = packageTreeArr.map(async element => {
+  element.dependencies = await populateDependencies(element, packageLock);
+});
+await Promise.all(m);
```

Comment:
Given that `m` isn't used anywhere else, you can just `await` the RHS of the previous statement.

Reply:
He still needs the `Promise.all`, which makes that previous line that much longer. I'm fine either way, but I see why he made the `m`.

Reply:
I came to the same conclusion, when I try to await the RHS my tests fail. So I am going to keep it unless there is a more readable way to write it out that is suggested

Reply:
I changed it so Promise.all is called on the return value of packageTreeArr.map